File: 연결리스트/leetcode876(Easy)-Middle of the Linked List.ts
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

// 런너 기법
function middleNode(head: ListNode | null): ListNode | null {
  if (!head) {
    return null
  }

  let fast = head
  let slow = fast

  while (fast.next && fast.next.next) {
    slow = slow.next
    fast = fast.next.next
  }

  return fast.next ? slow.next : slow
}

File: 연결리스트/leetcode876(Easy)-Middle of the Linked List.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

# 런너기법
class Solution:
    def middleNode(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if not head:
            return None

        slow = fast = head

        while fast.next and fast.next.next:
            slow = slow.next
            fast = fast.next.next

        if fast.next:
            return slow.next
        return slow